lib/hatchet/reaper.rb

```diff
@@ -1,167 +1,114 @@
-  #   another machine). When this happens, the system will automatically attempt to reap another app.
   class Reaper
     class AlreadyDeletedError < StandardError; end
 
     HATCHET_APP_LIMIT = Integer(ENV["HATCHET_APP_LIMIT"] || 20)  # the number of apps hatchet keeps around
     DEFAULT_REGEX = /^#{Regexp.escape(Hatchet::APP_PREFIX)}[a-f0-9]+/
+    TTL_MINUTES = ENV.fetch("HATCHET_ALIVE_TTL_MINUTES", "7").to_i
 
     attr_accessor :io, :hatchet_app_limit
 
     def initialize(api_rate_limit: , regex: DEFAULT_REGEX, io: STDOUT, hatchet_app_limit:  HATCHET_APP_LIMIT, initial_sleep: 10)
-      @api_rate_limit = api_rate_limit
-      @regex = regex
       @io = io
-      @finished_hatchet_apps = []
-      @unfinished_hatchet_apps = []
-      @app_count = 0
-      @hatchet_app_limit = hatchet_app_limit
+      @apps = []
+      @regex = regex
+      @limit = hatchet_app_limit
+      @api_rate_limit = api_rate_limit
       @reaper_throttle = ReaperThrottle.new(initial_sleep: initial_sleep)
     end
 
-    def cycle(app_exception_message: false)
+    # Called when we need an app, but are over limit or
+    # if an exception has occured that was possibly triggered
+    # by apps being over limit
+    def clean_old_or_sleep
       # Protect against parallel deletion of the same app on the same system
       mutex_file = File.open("#{Dir.tmpdir()}/hatchet_reaper_mutex", File::CREAT)
       mutex_file.flock(File::LOCK_EX)
 
-      refresh_app_list if @finished_hatchet_apps.empty?
+      destroy_older_apps(force_refresh: true)
 
-      # To be safe try to delete an app even if we're not over the limit
-      # since the exception may have been caused by going over the maximum account limit
-      if app_exception_message
+      if @apps.length > @limit
+        age = AppAge.new(created_at: @apps.last["created_at"], ttl_minutes: TTL_MINUTES)
+        @reaper_throttle.call(max_sleep: age.sleep_for_ttl) do |sleep_for|
           io.puts <<-EOM.strip_heredoc
-            WARNING: Running reaper due to exception on app
-                     #{stats_string}
-                     Exception: #{app_exception_message}
+            WARNING: Hatchet app limit reached (#{@apps.length}/#{@limit})
+                     All known apps are younger than #{TTL_MINUTES} minutes
           EOM
-        reap_once
-      end
 
-      while over_limit?
-        reap_once
+          sleep(sleep_for)
+        end
       end
     ensure
       mutex_file.close
     end
 
-    def stats_string
-      "total_app_count: #{@app_count}, hatchet_app_count: #{hatchet_app_count}/#{HATCHET_APP_LIMIT}, finished: #{@finished_hatchet_apps.length}, unfinished: #{@unfinished_hatchet_apps.length}"
-    end
-
-    def over_limit?
-      hatchet_app_count > hatchet_app_limit
+    # Destroys apps that are older than the given argument (expecting integer minutes)
+    def destroy_older_apps(minutes: TTL_MINUTES, force_refresh: @apps.empty?)
```

Comment:
There's no mutex on this method, which means if it's called directly (as happens on app startup), then all Hatchet processes will be spamming the Heroku API at once. In heroku-buildpack-python's case, this means 60 processes (well minus the one that's actually doing the delete) all getting 404 app not found, and draining the API rate limit.

I think the mutex needs to be moved here.

Also, it seems the destroyed apps never get removed from `@apps`?

Reply:
Example log showing 10,000 lines of attempts to delete apps, before I killed the CI job:
https://github.com/heroku/heroku-buildpack-python/actions/runs/4284904818/jobs/7462502820#step:5:10031